docs/user/reference/hook.md

```diff
@@ -137,6 +142,236 @@ The result is that the events are consumed in the order: `2-1-3`. Beware.
+* $JUJU_PEBBLE_CHECK_NAME holds the name of the Pebble check.
+
+## Hook ordering
+
+A charm's lifecycle consists of distinct **phases**:
+* install
+* operation
+* upgrade
+* tear down
+
+Generally, no assumptions can be made about order of hook execution. However, there are some limited guarantees
+about hook sequencing during install, upgrade and relation removal. 
+
+In normal operation, a unit will run at least the install, start, config-changed  and stop hooks over the course of its lifetime.
+
+### Install phase
+
+When a charm is first deployed, the following hook are executed in order before a charm reaches its operation phase:
+
+* storage-attached (if the unit has storage)
+* install
+* leader-elected (if the unit is a leader)
+* config-changed
+* start
+
+### Operation phase
+
+This phase is when a charm has completed its install operations and starts responding to events which correspond to interesting (relevant) changes to the Juju model.
+The behaviour is described in terms of hook kinds and is covered in subsequent sections.
+
+### Upgrade phase
+
+When a charm is upgraded, the `upgrade-charm` hook is usually followed by a `config-changed` hook.
+
+The `upgrade-charm` hook always runs once immediately after the charm directory
+contents have been changed by an unforced charm upgrade operation, and *may* do
+so after a forced upgrade; but will *not* be run after a forced upgrade from an
+existing error state. (Consequently, neither will the config-changed hook that
+would ordinarily follow the upgrade-charm.)
+
+
+### Tear down phase
+
+When an application is to be removed, the `stop` hook is the last hook to be run before the unit is destroyed.
+
+## Errors in hooks
+
+Hooks should ideally be idempotent, so that they can fail and be re-executed
+from scratch without trouble. As a hook author, you don't have complete control
+over the times your hook might be stopped: if the unit agent process is killed
+for any reason while running a hook, then when it recovers it will treat that
+hook as having failed -- just as if it had returned a non-zero exit code -- and
+request user intervention.
+
+It is unrealistic to expect great sophistication on the part of the average user,
+and as a charm author you should expect that users will attempt to re-execute
+failed hooks before attempting to investigate or understand the situation. You
+should therefore make every effort to ensure your hooks are idempotent when
+aborted and restarted.
+
+The most sophisticated charms will consider the nature of their operations with
+care, and will be prepared to internally retry any operations they suspect of
+having failed transiently, to ensure that they only request user intervention in
+the most trying circumstances; and will also be careful to log any relevant
+information or advice before signalling the error.
+
+## Hooks in detail
+
+This section describes the workflows and associated hook events which are important to the operation of Juju.
+
```

Comment:
The next title should be nested; so, ###.

The envvars it includes duplicate content from the envvar section above. I'd just give those kind-specific details here. Whatever you decided, let's not duplicate -- duplication is one of the biggest issues we've had with this content (because people would update it in one place but not the other etc.)

You said you were going to add further such nested section. Some of that content (for secrets and storage) is still in other docs. Let's move it from there to here. The relevant links are: 
- https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/secret/#secret-lifecycle 
- https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/storage/